src/udev/udev-manager.c

```diff
@@ -1135,118 +1135,104 @@ Manager* manager_new(void) {
-
-        assert(ret_ctrl);
-        assert(ret_netlink);
-
-        int n = sd_listen_fds_with_names(/* unset_environment = */ true, &names);
-        if (n < 0)
-                return n;
-
-        if (strv_length(names) != (size_t) n)
-                return -EIO;
-
-        for (int i = 0; i < n; i++) {
-                int fd = SD_LISTEN_FDS_START + i;
-
-                if (streq(names[i], "varlink"))
-                        continue; /* The fd will be handled by sd_varlink_server_listen_auto(). */
-
-                if (sd_is_socket(fd, AF_UNIX, SOCK_SEQPACKET, -1) > 0) {
-                        if (ctrl_fd >= 0) {
-                                log_debug("Received multiple seqpacket socket (%s), ignoring.", names[i]);
-                                goto unused;
-                        }
-
-                        ctrl_fd = fd;
-                        continue;
-                }
+static int manager_init_ctrl(Manager *manager, int fd) {
+        int r;
 
-                if (sd_is_socket(fd, AF_NETLINK, SOCK_RAW, -1) > 0) {
-                        if (netlink_fd >= 0) {
-                                log_debug("Received multiple netlink socket (%s), ignoring.", names[i]);
-                                goto unused;
-                        }
+        assert(manager);
 
-                        netlink_fd = fd;
-                        continue;
-                }
+        /* This consumes passed file descriptor on success. */
 
-                log_debug("Received unexpected fd (%s), ignoring.", names[i]);
+        if (fd >= 0) {
+                if (manager->ctrl)
+                        return log_warning_errno(SYNTHETIC_ERRNO(EALREADY), "Received multiple control socket (%i), ignoring.", fd);
 
-        unused:
-                close_and_notify_warn(fd, names[i]);
+                r = sd_is_socket(fd, AF_UNIX, SOCK_SEQPACKET, -1);
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to verify socket type of %i, ignoring: %m", fd);
+                if (r == 0)
+                        return log_warning_errno(SYNTHETIC_ERRNO(EINVAL), "Received invalid control socket (%i), ignoring.", fd);
+        } else {
+                if (manager->ctrl)
+                        return 0;
         }
 
-        *ret_ctrl = TAKE_FD(ctrl_fd);
-        *ret_netlink = TAKE_FD(netlink_fd);
+        r = udev_ctrl_new_from_fd(&manager->ctrl, fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to initialize udev control socket: %m");
+
         return 0;
 }
 
-static int manager_init_ctrl(Manager *manager, int fd_ctrl) {
-        _cleanup_(udev_ctrl_unrefp) UdevCtrl *ctrl = NULL;
-        _cleanup_close_ int fd = fd_ctrl;
+static int manager_init_device_monitor(Manager *manager, int fd) {
         int r;
 
         assert(manager);
 
-        /* This consumes passed file descriptor. */
+        /* This consumes passed file descriptor on success. */
```

Comment:
also here. the fd cleanup got dropped?

Reply:
I see that the caller will close_and_notify_warn(). But please update the comments.